pydantic/_internal/_std_types_schema.py

```diff
@@ -83,29 +91,37 @@ def to_enum(__input_value: Any, _: core_schema.ValidationInfo) -> Enum:
     )
 
     lax: CoreSchema
-    json_type: Literal['int', 'float', 'str']
+    json_types: set[Literal['int', 'float', 'str']]
     if issubclass(enum_type, int):
         # this handles `IntEnum`, and also `Foobar(int, Enum)`
         updates['type'] = 'integer'
         lax = core_schema.chain_schema(
             [core_schema.int_schema(), literal_schema, core_schema.general_plain_validator_function(to_enum)],
-            metadata=metadata,
         )
-        json_type = 'int'
+        json_types = {'int'}
     elif issubclass(enum_type, str):
         # this handles `StrEnum` (3.11 only), and also `Foobar(str, Enum)`
         updates['type'] = 'string'
         lax = core_schema.chain_schema(
             [core_schema.str_schema(), literal_schema, core_schema.general_plain_validator_function(to_enum)],
-            metadata=metadata,
         )
-        json_type = 'str'
+        json_types = {'str'}
     else:
-        lax = core_schema.general_after_validator_function(to_enum, literal_schema, metadata=metadata)
-        json_type = 'str'
+
+        def wrap_to_enum(
+            __input_value: Any, __handler: ValidatorFunctionWrapHandler, _: core_schema.ValidationInfo
+        ) -> Enum:
+            if isinstance(__input_value, enum_type):
+                __input_value = __input_value.value
+            value = __handler(__input_value)
```

Comment:
What would `__handler` ever represent? Is there a test I can look at that demonstrates why this needs to be a wrap handler and not an `isinstance` handler or something?

My understanding is that we want `class MyEnum({int,str,float}, Enum)` to work w/ JSON but anything else would work w/ Python but not JSON.

Another thing is that I think that Enums should ~ always be strict:

```python
from enum import Enum, auto

class MyEnum2(Enum):
    x = auto()

AnalyzedType(MyEnum1).validate_python(1)  # fails

class MyEnum2(Enum):
    x = 1

AnalyzedType(MyEnum2).validate_python(1)  # fails
```

But that just be my opinion.

Reply:
I was using `__handler` to get it to do the check that the value is in the allowed literal values. What's happening here is that it is unpacking the enum into its value, validating the value, then repackaging it into the Enum.

I think the reason this was necessary now was because the literal thing was using values and not Enums. Maybe that just doesn't need to be the case.

Also I realized you are right it should probably just always be strict. I'll try reworking this.

Reply:
Making enum parsing always strict is a significant breaking change so I'm not going to do that here, but I would be happy to make that change at some point if we establish broader consensus.